File: C++/992. Subarrays with K Different Integers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        return countK(nums, k) - countK(nums, k - 1);
    }    
private:
    int countK(vector<int>& nums, int k) {
        unordered_map<int, int> counter;
        int left = 0;
        int distinct_count = 0;
        int result = 0;
        for (int right = 0; right < nums.size(); ++right) {
            if (counter[nums[right]] == 0) {
                distinct_count++;
            }
            counter[nums[right]]++;
            while (distinct_count > k) {
                counter[nums[left]]--;
                if (counter[nums[left]] == 0) {
                    distinct_count--;
                }
                left++;
            }
            result += right - left + 1;
        }
        return result;
    }
};

File: JAVA/992. Subarrays with K Different Integers.java
class Solution {
    public static int atmost(int nums[],int k){
        HashMap<Integer,Integer> map=new HashMap<>();
        int count=0;
        int i=0;
        for(int j=0;j<nums.length;j++){
            map.put(nums[j],map.getOrDefault(nums[j],0)+1);
            while(map.size()>k){
                map.put(nums[i],map.get(nums[i])-1);
                if(map.get(nums[i])==0){
                    map.remove(nums[i]);
                }
                i++;
            }
            count=count+(j-i+1);
        }
        return count;
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        int atmostk=atmost(nums,k);
        int atmostk1=atmost(nums,k-1);
        return atmostk-atmostk1;
    }
}
